src/driver/op.rs

```diff
@@ -147,46 +174,87 @@ where
                 me.index = usize::MAX;
                 Poll::Ready(me.data.take().unwrap().complete(cqe))
             }
+            Lifecycle::CompletionList(..) => {
+                unreachable!()
+            }
         }
     }
 }
 
-impl<T> Drop for Op<T> {
+impl<T, CqeType> Drop for Op<T, CqeType> {
     fn drop(&mut self) {
+        use std::mem;
+
         let mut inner = self.driver.borrow_mut();
-        let lifecycle = match inner.ops.get_mut(self.index) {
-            Some(lifecycle) => lifecycle,
+        let (lifecycle, completions) = match inner.ops.get_mut(self.index) {
+            Some(val) => val,
             None => return,
         };
 
-        match lifecycle {
+        match mem::replace(lifecycle, Lifecycle::Submitted) {
             Lifecycle::Submitted | Lifecycle::Waiting(_) => {
                 *lifecycle = Lifecycle::Ignored(Box::new(self.data.take()));
             }
             Lifecycle::Completed(..) => {
                 inner.ops.remove(self.index);
             }
+            Lifecycle::CompletionList(indices) => {
+                // Deallocate list entries, recording if the more CQE's are expected
+                let more = {
+                    let mut list = indices.into_list(completions);
+                    io_uring::cqueue::more(list.peek_end().unwrap().flags)
+                    // Dropping list dealloctes the list entries
+                };
+                if more {
+                    // If more are expected, we have to keep the op around
+                    *lifecycle = Lifecycle::Ignored(Box::new(self.data.take()));
+                } else {
+                    inner.ops.remove(self.index);
+                }
+            }
             Lifecycle::Ignored(..) => unreachable!(),
         }
     }
 }
 
 impl Lifecycle {
-    pub(super) fn complete(&mut self, cqe: CqeResult) -> bool {
+    pub(super) fn complete(&mut self, completions: &mut Slab<Completion>, cqe: CqeResult) -> bool {
         use std::mem;
 
         match mem::replace(self, Lifecycle::Submitted) {
-            Lifecycle::Submitted => {
-                *self = Lifecycle::Completed(cqe);
+            x @ Lifecycle::Submitted | x @ Lifecycle::Waiting(..) => {
+                if io_uring::cqueue::more(cqe.flags) {
+                    let mut list = SlabListIndices::new().into_list(completions);
+                    list.push(cqe);
+                    *self = Lifecycle::CompletionList(list.into_indices());
+                } else {
+                    *self = Lifecycle::Completed(cqe);
+                }
+                if let Lifecycle::Waiting(waker) = x {
+                    waker.wake();
+                }
```

Comment:
There are 2 design decisions here, 

- completions with more flag set go straight onto the slab. It would be equally valid to put them into the Completed variant, and only upgrade to the slab if another is received before Poll.

- wake is called whenever we have more work. It would also be possible to defer calling wake until a completion without more is received. This behavior could be configured per Op type.

Reply:
Those are both optimisations for things we don't have yet, so I think definitely another PR